test/integration/http2_integration_test.h

```diff
@@ -115,17 +111,32 @@ class Http2FloodMitigationTest : public Http2FrameIntegrationTest {
     bool writev_returns_egain_ ABSL_GUARDED_BY(mutex_) = false;
   };
 
+  SocketInterfaceSwap();
+  ~SocketInterfaceSwap();
+
+protected:
+  Envoy::Network::SocketInterface* const previous_socket_interface_{
+      Envoy::Network::SocketInterfaceSingleton::getExisting()};
+  std::shared_ptr<IoHandleMatcher> writev_matcher_{std::make_shared<IoHandleMatcher>()};
+  std::unique_ptr<Envoy::Network::SocketInterfaceLoader> test_socket_interface_loader_;
+};
+
+// It is important that the new socket interface is installed before any I/O activity starts and
+// the previous one is restored after all I/O activity stops. Since the HttpIntegrationTest
+// destructor stops Envoy the SocketInterfaceSwap destructor needs to run after it. This order of
+// multiple inheritance ensures that SocketInterfaceSwap destructor runs after
+// Http2FrameIntegrationTest destructor completes.
+class Http2FloodMitigationTest : public SocketInterfaceSwap, public Http2FrameIntegrationTest {
```

Comment:
This feels somewhat fragile.  While the comments are huge help, I have a question you probably have the context to answer, which is if this needs to be done per test run, or if it can be done in SetUp/TearDownTestSuite

Reply:
It needs to be done per test run because the test needs to override socket API for specific sockets. Which is what the `writev_matcher_` is for and it is test specific and is stored in the socket interface singleton. The requirement is to reset the singleton after the call to the `cleanupUpstreamAndDownstream()`, which this approach guarantees. We could also make `cleanupUpstreamAndDownstream()` idempotent so it can be called multiple times.